databricks/sdk/_widgets/__init__.py

```diff
@@ -0,0 +1,76 @@
+import logging
+import typing
+import warnings
+from abc import ABC, abstractmethod
+
+
+class WidgetUtils(ABC):
+
+    @property
+    @abstractmethod
+    def _widget_names(self) -> typing.List[str]:
+        pass
+
+    def get(self, name: str):
+        return self._get(name)
+
+    @abstractmethod
+    def _get(self, name: str) -> str:
+        pass
+
+    def getArgument(self, name: str, default_value: typing.Optional[str] = None):
+        try:
+            return self.get(name)
+        except Exception:
+            return default_value
+
+    def remove(self, name: str):
+        self._remove(name)
+
+    @abstractmethod
+    def _remove(self, name: str):
+        pass
+
+    def removeAll(self):
+        self._remove_all()
+
+    @abstractmethod
+    def _remove_all(self):
+        pass
+
+
+try:
+    # We only use ipywidgets if we are in a notebook interactive shell otherwise we raise error,
+    # to fallback to using default_widgets. Also, users WILL have IPython in their notebooks (jupyter),
+    # because we DO NOT SUPPORT any other notebook backends, and hence fallback to default_widgets.
+    from IPython.core.getipython import get_ipython
+
+    # Iterate over the mro of the current ipython instance, to find ZMQInteractiveShell (jupyter). When used
+    # from REPL or file, this check will fail, since the mro only contains TerminalInteractiveShell.
+    if len(list(filter(lambda i: i.__name__ == 'ZMQInteractiveShell', get_ipython().__class__.__mro__))) == 0:
+        logging.debug("Not in an interactive notebook. Skipping ipywidgets implementation for dbutils.")
+        raise EnvironmentError("Not in an interactive notebook.")
```

Comment:
Not a blocker for this change, but have we thought about how widgets would work for users running programs using the Databricks SDK locally? Current fallback implementation uses the default value only, but perhaps that implementation could look for an environment variable named `DB_WIDGET_<NAME>` to allow automated use-cases to set widget values.

Reply:
Let me think about this offline and we can make this a separate PR. This PR is already getting complicated.